123-job-jwt/proposal.md

```diff
@@ -0,0 +1,137 @@
+# Summary
+
+Mint a signed JSON Web Token (JWT) and deliver to the job runtime container.
+Additionally, provide a validation endpoint to allow for external or internal
+validation of the issued JWT.
+
+## Motivation
+
+Concourse runners lack an identity mechanism that can be used to authenticate to
+external systems. This lack means that an initial secret must be provided to
+each pipeline in Concourse in a manual manner. Many other CI solutions and cloud
+providers solve the "secret zero" problem by provisioning a signed JWT to their
+workloads automatically, and providing a path to signature verification in a
+manner that the external applications can trust.
+
+Should this be implemented, Concourse jobs will be able to leverage this
+automatically generated JWT to be the "secret zero" needed to communicate with
+external systems (such as Hashicorp Vault, AWS, and others) and will no longer
+require a Concourse administrator (or person configuring a pipeline) to provide
+a static credential to Concourse. This alleviates the need for a human to handle
+sensitive credentials, increases security by ensuring credentials are not
+statically configured and JWTs are ephemeral to each run, and reduces
+Concourse's need to properly handle and encrypt data (can be limited to storing
+the JWT private signing key).
```

Comment:
That key should probably be created by concourse, and treated as some state surviving reboots, but be rotated periodically: https://auth0.com/docs/get-started/tenant-settings/signing-keys/rotate-signing-keys#use-the-dashboard

Reply:
Absolutely, there could be a process for manually supplying a key for testing environments, but that kind of pattern should be avoided for production environments, and by default it should simply generate its own key (perhaps on every reboot, perhaps also on a schedule) without any effort or consideration from the Concourse deployment standpoint.

Reply:
Yes, my point here though was that old keys still need to be returned for a while in the JWKS endpoint, so treating it entirely as runtime-only state won't work.

Periodic rotation, without any need for the admin to do anything should do the default. If the key material gets compromised, there should be a way to revoke the key immediately - similar to how the screenshots on the link shared above.